### PR TITLE
feat: forester: add state batched operations

### PR DESCRIPTION
**Pull request by sergeytimoshin**

### CODE REVIEW

File: test-utils/src/indexer/test_indexer.rs [hunk 1]
```diff
@@ -139,6 +142,30 @@ impl<R: RpcConnection + Send + Sync + 'static> Indexer<R> for TestIndexer<R> {
         Err(IndexerError::Custom("Merkle tree not found".to_string()))
     }
 
+    async fn get_subtrees(
```
Comment (ananas-block) [inline test-utils/src/indexer/test_indexer.rs:145]:
```suggestion
    async fn get_sub_trees(
```
feels more intuitive tbh

File: forester/src/batched_ops.rs [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+        //     let output_queue = ZeroCopyBatchedQueueAccount::from_bytes_mut(
+        //         output_queue_account.data.as_mut_slice(),
+        //     )
+        //     .unwrap();
+        //     let idx = output_queue.get_account().queue.next_full_batch_index;
+        //     let batch = &output_queue.batches[idx as usize];
+        //     println!("is_batch_ready: batch: {:?}", batch);
+        //     batch.get_state() == BatchState::ReadyToUpdateTree
+        // };
+        // is_batch_ready
```
Comment (ananas-block) [inline forester/src/batched_ops.rs:63]: leftover?

File: forester-utils/src/indexer/mod.rs [hunk 3]
```diff
@@ -217,6 +239,25 @@ pub trait Indexer<R: RpcConnection>: Sync + Send + Debug + 'static {
...
+        &mut self,
+        _rpc: &mut R,
+        _merkle_tree_pubkey: Pubkey,
+        _output_queue_pubkey: Pubkey,
+        _num_inserted_zkps: u64,
+    ) {
+        unimplemented!()
+    }
+
+    async fn update_test_indexer_after_nullification(
```
Comment (ananas-block) [inline forester-utils/src/indexer/mod.rs:253]: these should be part of test indexer not the indexer trait right?

File: client/src/rpc/solana_rpc.rs [hunk 4]
```diff
@@ -441,4 +467,24 @@ impl RpcConnection for SolanaRpcConnection {
...
     }
+
+    async fn get_transaction_slot(&mut self, signature: &Signature) -> Result<u64, RpcError> {
+        self.retry(|| async {
+            Ok(self
+                .client
+                .get_transaction_with_config(
+                    signature,
+                    // UiTransactionEncoding::Json,
+                    // self.client.commitment(),
```
Comment (ananas-block) [inline client/src/rpc/solana_rpc.rs:478]: leftover?

File: client/src/rpc/solana_rpc.rs [hunk 5]
```diff
@@ -318,24 +318,50 @@ impl RpcConnection for SolanaRpcConnection {
...
+                println!("network_fee: {}", network_fee);
+                println!("num signers {}", deduped_signers.len());
+                // return Err(RpcError::from(BanksClientError::TransactionError(
+                //     TransactionError::InstructionError(0, InstructionError::Custom(11111)),
+                // )));
                 return Err(RpcError::AssertRpcError(format!("unexpected balance after transaction: expected {expected_post_balance}, got {post_balance}")));
             }
+            // if post_balance as i64 != expected_post_balance {
+            //     return Err(RpcError::AssertRpcError(format!("unexpected balance after transaction: expected {expected_post_balance}, got {post_balance}")));
+            // }
```
Comment (ananas-block) [inline client/src/rpc/solana_rpc.rs:364]: leftover?

Comment (ananas-block) [reply client/src/rpc/solana_rpc.rs:364]: and prints above

File: forester-utils/src/indexer/mod.rs (hunk 3 above)
Comment (sergeytimoshin) [reply forester-utils/src/indexer/mod.rs:253]: Yes, I'll move it there in the test_indexer's PR

File: test-utils/src/indexer/test_indexer.rs (hunk 1 above)
Comment (sergeytimoshin) [reply test-utils/src/indexer/test_indexer.rs:145]: hm why? "subtree" is a standard term in graph theory

File: forester/src/batched_ops.rs [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+    async fn is_batch_ready(&self) -> bool {
+        let mut rpc = self.rpc_pool.get_connection().await.unwrap();
+        let is_batch_ready = {
+            let mut output_queue_account =
+                rpc.get_account(self.output_queue).await.unwrap().unwrap();
+            let output_queue = BatchedQueueAccount::output_queue_from_bytes_mut(
+                output_queue_account.data.as_mut_slice(),
+            )
+            .unwrap();
+            output_queue.get_batch_num_inserted_in_current_batch() > 0
```
Comment (ananas-block) [inline forester/src/batched_ops.rs:58]: I don't think this is a robust way to detect whether a batch or a part of a batch is ready for insertion.
This is the logic we use to check update readiness onchain, you can either reuse it and catch an error here or add a sub method which returns a boolean.
https://github.com/Lightprotocol/light-protocol/blob/3f44fbf094b64ccd82751bf9bbba84f0be7969d4/merkle-tree/batched/src/batch.rs#L114

Comment (ananas-block) [reply forester/src/batched_ops.rs:58]: alternatively you can probably do:
```suggestion
            output_queue.get_batch_num_inserted_in_current_batch() % output_queue.zkp_batch_size == 0
```

**Review by ananas-block — APPROVED** [review]

**Pull request merged**